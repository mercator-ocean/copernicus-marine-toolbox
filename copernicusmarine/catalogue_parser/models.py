import re
from dataclasses import dataclass
from enum import Enum
from typing import Optional, Type, TypeVar, Union

import pystac

from copernicusmarine.command_line_interface.exception_handler import (
    log_exception_debug,
)
from copernicusmarine.core_functions.utils import (
    datetime_parser,
    next_or_raise_exception,
)

# Output Types definitions

VERSION_DEFAULT = "default"
PART_DEFAULT = "default"


# Service types
class _ServiceName(str, Enum):
    GEOSERIES = "arco-geo-series"
    TIMESERIES = "arco-time-series"
    FILES = "original-files"
    WMTS = "wmts"
    OMI_ARCO = "omi-arco"
    STATIC_ARCO = "static-arco"


class _ServiceShortName(str, Enum):
    GEOSERIES = "geoseries"
    TIMESERIES = "timeseries"
    FILES = "files"
    WMTS = "wmts"
    OMI_ARCO = "omi-arco"
    STATIC_ARCO = "static-arco"


@dataclass(frozen=True)
class _Service:
    service_name: _ServiceName
    short_name: _ServiceShortName

    def aliases(self) -> list[str]:
        return (
            [self.service_name.value, self.short_name.value]
            if self.short_name.value != self.service_name.value
            else [self.service_name.value]
        )

    def to_json_dict(self):
        return {
            "service_name": self.service_name.value,
            "short_name": self.short_name.value,
        }


class CopernicusMarineDatasetServiceType(_Service, Enum):
    GEOSERIES = _ServiceName.GEOSERIES, _ServiceShortName.GEOSERIES
    TIMESERIES = (
        _ServiceName.TIMESERIES,
        _ServiceShortName.TIMESERIES,
    )
    FILES = _ServiceName.FILES, _ServiceShortName.FILES
    WMTS = _ServiceName.WMTS, _ServiceShortName.WMTS
    OMI_ARCO = _ServiceName.OMI_ARCO, _ServiceShortName.OMI_ARCO
    STATIC_ARCO = _ServiceName.STATIC_ARCO, _ServiceShortName.STATIC_ARCO


def _service_type_from_web_api_string(
    name: str,
) -> CopernicusMarineDatasetServiceType:
    class WebApi(Enum):
        GEOSERIES = "timeChunked"
        TIMESERIES = "geoChunked"
        FILES = "native"
        WMTS = "wmts"
        OMI_ARCO = "omi"
        STATIC_ARCO = "static"

    web_api_mapping = {
        WebApi.GEOSERIES: CopernicusMarineDatasetServiceType.GEOSERIES,
        WebApi.TIMESERIES: CopernicusMarineDatasetServiceType.TIMESERIES,
        WebApi.FILES: CopernicusMarineDatasetServiceType.FILES,
        WebApi.WMTS: CopernicusMarineDatasetServiceType.WMTS,
        WebApi.OMI_ARCO: CopernicusMarineDatasetServiceType.OMI_ARCO,
        WebApi.STATIC_ARCO: CopernicusMarineDatasetServiceType.STATIC_ARCO,
    }

    return next_or_raise_exception(
        (
            service_type
            for service_web_api, service_type in web_api_mapping.items()
            if service_web_api.value == name
        ),
        ServiceNotHandled(name),
    )


class ServiceNotHandled(Exception):
    ...


# service formats
class CopernicusMarineServiceFormat(str, Enum):
    ZARR = "zarr"
    SQLITE = "sqlite"


@dataclass
class CopernicusMarineCoordinate:
    coordinates_id: str
    units: str
    minimum_value: Optional[float]
    maximum_value: Optional[float]
    step: Optional[float]
    values: Optional[list[Union[float, int]]]
    chunking_length: Optional[int]
    chunk_type: Optional[str]
    chunk_reference_coordinate: Optional[int]
    chunk_geometric_factor: Optional[int]

    Coordinate = TypeVar("Coordinate", bound="CopernicusMarineCoordinate")

    @classmethod
    def from_metadata_item(
        cls: Type[Coordinate],
        variable_id: str,
        dimension: str,
        dimension_metadata: dict,
        arco_data_metadata_producer_valid_start_date: Optional[str],
    ) -> Coordinate:
        coordinates_info = dimension_metadata.get("coords", {})
        if (
            arco_data_metadata_producer_valid_start_date
            and dimension == "time"
        ):
            minimum_value = (
                CopernicusMarineCoordinate._format_admp_valid_start_date(
                    arco_data_metadata_producer_valid_start_date,
                    to_timestamp=isinstance(coordinates_info.get("min"), int),
                )
            )
        else:
            minimum_value = coordinates_info.get("min")
        chunking_length = dimension_metadata.get("chunkLen")
        if isinstance(chunking_length, dict):
            chunking_length = chunking_length.get(variable_id)

        coordinate = cls(
            coordinates_id=dimension,
            units=dimension_metadata.get("units") or "",
            minimum_value=minimum_value,  # type: ignore
            maximum_value=coordinates_info.get("max"),
            step=coordinates_info.get("step"),
            values=coordinates_info.get("values"),
            chunking_length=chunking_length,
            chunk_type=dimension_metadata.get("chunkType"),
            chunk_reference_coordinate=dimension_metadata.get("chunkRefCoord"),
            chunk_geometric_factor=dimension_metadata.get(
                "chunkGeometricFactor", {}
            ).get(variable_id),
        )
        if dimension == "elevation":
            coordinate._convert_elevation_to_depth()
        return coordinate

    @staticmethod
    def _format_admp_valid_start_date(
        arco_data_metadata_producer_valid_start_date: str,
        to_timestamp: bool = False,
    ) -> Union[str, int]:
        if to_timestamp:
            return int(
                datetime_parser(
                    arco_data_metadata_producer_valid_start_date.split(".")[0]
                ).timestamp()
                * 1000
            )
        return arco_data_metadata_producer_valid_start_date.split(".")[0]

    def _convert_elevation_to_depth(self):
        self.coordinates_id = "depth"
        minimum_elevation = self.minimum_value
        maximum_elevation = self.maximum_value
        if minimum_elevation is not None:
            self.maximum_value = -minimum_elevation
        else:
            self.maximum_value = None
        if maximum_elevation is not None:
            self.minimum_value = -maximum_elevation
        else:
            self.minimum_value = None
        if self.values is not None:
            self.values = [-value for value in self.values]


@dataclass
class CopernicusMarineVariable:
    short_name: str
    standard_name: str
    units: str
    bbox: Optional[list[float]]
    coordinates: list[CopernicusMarineCoordinate]

    Variable = TypeVar("Variable", bound="CopernicusMarineVariable")

    @classmethod
    def from_metadata_item(
        cls: Type[Variable],
        metadata_item: pystac.Item,
        asset: pystac.Asset,
        variable_id: str,
        bbox: Optional[list[float]],
    ) -> Variable:
        cube_variables = metadata_item.properties["cube:variables"]
        cube_variable = cube_variables[variable_id]

        # to get coordinates
        extra_fields_asset = asset.extra_fields
        dimensions = extra_fields_asset.get("viewDims") or {}
        return cls(
            short_name=variable_id,
            standard_name=cube_variable["standardName"],
            units=cube_variable.get("unit") or "",
            bbox=bbox,
            coordinates=[
                CopernicusMarineCoordinate.from_metadata_item(
                    variable_id,
                    dimension,
                    dimension_metadata,
                    metadata_item.properties.get("admp_valid_start_date"),
                )
                for dimension, dimension_metadata in dimensions.items()
            ],
        )


@dataclass
class CopernicusMarineService:
    service_type: CopernicusMarineDatasetServiceType
    service_format: Optional[CopernicusMarineServiceFormat]
    uri: str
    variables: list[CopernicusMarineVariable]

    Service = TypeVar("Service", bound="CopernicusMarineService")

    @classmethod
    def from_metadata_item(
        cls: Type[Service],
        metadata_item: pystac.Item,
        service_name: str,
        asset: pystac.Asset,
    ) -> Optional[Service]:
        try:
            service_uri = asset.get_absolute_href()
            if not service_uri:
                raise ServiceNotHandled(service_name)
            service_type = _service_type_from_web_api_string(service_name)
            service_format = None
            admp_in_preparation = metadata_item.properties.get(
                "admp_in_preparation"
            )
            if asset.media_type and "zarr" in asset.media_type:
                service_format = CopernicusMarineServiceFormat.ZARR
            elif asset.media_type and "sqlite3" in asset.media_type:
                service_format = CopernicusMarineServiceFormat.SQLITE

            if not service_uri.endswith("/"):
                if admp_in_preparation and (
                    service_type
                    == CopernicusMarineDatasetServiceType.GEOSERIES
                    or service_type
                    == CopernicusMarineDatasetServiceType.TIMESERIES
                ):
                    return None
                else:
                    bbox = metadata_item.bbox
                    return cls(
                        service_type=service_type,
                        uri=service_uri,
                        variables=[
                            CopernicusMarineVariable.from_metadata_item(
                                metadata_item, asset, var_cube["id"], bbox
                            )
                            for var_cube in metadata_item.properties[
                                "cube:variables"
                            ].values()
                        ],
                        service_format=service_format,
                    )
            return None
        except ServiceNotHandled as service_not_handled:
            log_exception_debug(service_not_handled)
            return None


@dataclass
class CopernicusMarineVersionPart:
    name: str
    services: list[CopernicusMarineService]
    retired_date: Optional[str]
    released_date: Optional[str]

    VersionPart = TypeVar("VersionPart", bound="CopernicusMarineVersionPart")

    @classmethod
    def from_metadata_item(
        cls: Type[VersionPart], metadata_item: pystac.Item, part_name: str
    ) -> Optional[VersionPart]:
        retired_date = metadata_item.properties.get("admp_retired_date")
        released_date = metadata_item.properties.get("admp_released_date")
        if retired_date and datetime_parser(retired_date) < datetime_parser(
            "now"
        ):
            return None
        services = [
            service
            for metadata_service_name, asset in metadata_item.get_assets().items()
            if (
                service := CopernicusMarineService.from_metadata_item(
                    metadata_item,
                    metadata_service_name,
                    asset,
                )
            )
        ]
        if not services:
            return None
        services = services
        return cls(
            name=part_name,
            services=services,
            retired_date=retired_date,
            released_date=released_date,
        )

    def get_service_by_service_type(
        self, service_type: CopernicusMarineDatasetServiceType
    ):
        return next(
            service
            for service in self.services
            if service.service_type == service_type
        )


@dataclass
class CopernicusMarineDatasetVersion:
    label: str
    parts: list[CopernicusMarineVersionPart]

    def get_part(
        self, force_part: Optional[str]
    ) -> CopernicusMarineVersionPart:
        wanted_part = force_part or PART_DEFAULT
        for part in self.parts:
            if part.name == wanted_part:
                return part
            elif not force_part:
                return part
        raise dataset_version_part_not_found_exception(self)

    def sort_parts(self) -> tuple[Optional[str], Optional[str]]:
        not_released_parts = {
            part.name
            for part in self.parts
            if part.released_date
            and datetime_parser(part.released_date) > datetime_parser("now")
        }
        will_be_retired_parts = {
            part.name: datetime_parser(part.retired_date).timestamp()
            for part in self.parts
            if part.retired_date
        }
        max_retired_timestamp = 0.0
        if will_be_retired_parts:
            max_retired_timestamp = max(will_be_retired_parts.values()) + 1
        self.parts = sorted(
            self.parts,
            key=lambda x: (
                x.name in not_released_parts,
                max_retired_timestamp
                - will_be_retired_parts.get(x.name, max_retired_timestamp),
                -(x.name == PART_DEFAULT),
                -(x.name == "latest"),  # for INSITU datasets
                -(x.name == "bathy"),  # for STATIC datasets
                x.name,
            ),
        )
        return self.parts[0].released_date, self.parts[0].retired_date


@dataclass
class CopernicusMarineProductDataset:
    dataset_id: str
    dataset_name: str
    versions: list[CopernicusMarineDatasetVersion]

    def get_version(
        self, force_version: Optional[str]
    ) -> CopernicusMarineDatasetVersion:
        wanted_version = force_version or VERSION_DEFAULT
        for version in self.versions:
            if version.label == wanted_version:
                return version
            elif not force_version:
                return version
        raise dataset_version_not_found_exception(self)

    def sort_versions(self) -> None:
        not_released_versions: set[str] = set()
        retired_dates = {}
        for version in self.versions:
            released_date, retired_date = version.sort_parts()
            if released_date and datetime_parser(
                released_date
            ) > datetime_parser("now"):
                not_released_versions.add(version.label)
            if retired_date:
                retired_dates[version.label] = retired_date

        self.versions = sorted(
            self.versions,
            key=lambda x: (
                -(x.label in not_released_versions),
                retired_dates.get(x.label, "9999-12-31"),
                -(x.label == VERSION_DEFAULT),
                x.label,
            ),
            reverse=True,
        )

    def parse_dataset_metadata_items(
        self, metadata_items: list[pystac.Item]
    ) -> None:
        all_versions = set()
        for metadata_item in metadata_items:
            (
                _,
                dataset_version,
                dataset_part,
            ) = get_version_and_part_from_full_dataset_id(metadata_item.id)
            part = CopernicusMarineVersionPart.from_metadata_item(
                metadata_item, dataset_part
            )
            if not part:
                continue
            if dataset_version in all_versions:
                for version in self.versions:
                    if version.label == dataset_version:
                        version.parts.append(part)
                        break
            else:
                all_versions.add(dataset_version)
                version = CopernicusMarineDatasetVersion(
                    label=dataset_version, parts=[part]
                )
                self.versions.append(version)


@dataclass
class CopernicusMarineProduct:
    title: str
    product_id: str
    thumbnail_url: str
    description: str
    digital_object_identifier: Optional[str]
    sources: list[str]
    processing_level: Optional[str]
    production_center: str
    keywords: Optional[list[str]]
    datasets: list[CopernicusMarineProductDataset]


@dataclass
class CopernicusMarineCatalogue:
    products: list[CopernicusMarineProduct]

    def filter_only_official_versions_and_parts(self):
        products_to_remove = []
        for product in self.products:
            datasets_to_remove = []
            for dataset in product.datasets:
                latest_version = dataset.versions[0]
                parts_to_remove = []
                for part in latest_version.parts:
                    if part.released_date and datetime_parser(
                        part.released_date
                    ) > datetime_parser("now"):
                        parts_to_remove.append(part)
                for part_to_remove in parts_to_remove:
                    latest_version.parts.remove(part_to_remove)
                if not latest_version.parts:
                    datasets_to_remove.append(dataset)
                else:
                    dataset.versions = [latest_version]
            for dataset_to_remove in datasets_to_remove:
                product.datasets.remove(dataset_to_remove)
            if not product.datasets:
                products_to_remove.append(product)
        for product_to_remove in products_to_remove:
            self.products.remove(product_to_remove)


# Errors
class DatasetVersionPartNotFound(Exception):
    ...


class DatasetVersionNotFound(Exception):
    ...


def dataset_version_part_not_found_exception(
    version: CopernicusMarineDatasetVersion,
) -> DatasetVersionPartNotFound:
    return DatasetVersionPartNotFound(
        f"No part found for version {version.label}"
    )


def dataset_version_not_found_exception(
    dataset: CopernicusMarineProductDataset,
) -> DatasetVersionNotFound:
    return DatasetVersionNotFound(
        f"No version found for dataset {dataset.dataset_id}"
    )


REGEX_PATTERN_DATE_YYYYMM = r"[12]\d{3}(0[1-9]|1[0-2])"
PART_SEPARATOR = "--ext--"


def get_version_and_part_from_full_dataset_id(
    full_dataset_id: str,
) -> tuple[str, str, str]:
    if PART_SEPARATOR in full_dataset_id:
        name_with_maybe_version, part = full_dataset_id.split(PART_SEPARATOR)
    else:
        name_with_maybe_version = full_dataset_id
        part = PART_DEFAULT
    pattern = rf"^(.*?)(?:_({REGEX_PATTERN_DATE_YYYYMM}))?$"
    match = re.match(pattern, name_with_maybe_version)
    if match:
        dataset_name = match.group(1)
        version = match.group(2) or VERSION_DEFAULT
    else:
        raise Exception(f"Could not parse dataset id: {full_dataset_id}")
    return dataset_name, version, part
