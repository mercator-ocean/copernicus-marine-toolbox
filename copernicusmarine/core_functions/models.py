from typing import Literal, get_args

FileFormat = Literal["netcdf", "zarr"]
DEFAULT_FILE_FORMAT: FileFormat = "netcdf"
DEFAULT_FILE_FORMATS = list(get_args(FileFormat))

FileExtension = Literal[".nc", ".zarr"]
DEFAULT_FILE_EXTENSION: FileExtension = ".nc"
DEFAULT_FILE_EXTENSIONS = list(get_args(FileExtension))

SubsetMethod = Literal["nearest", "strict"]
DEFAULT_SUBSET_METHOD: SubsetMethod = "nearest"
DEFAULT_SUBSET_METHODS = list(get_args(SubsetMethod))

BoundingBoxMethod = Literal["inside", "outside"]
DEFAULT_BOUNDING_BOX_METHOD: BoundingBoxMethod = "inside"
DEFAULT_BOUNDING_BOX_METHODS = list(get_args(BoundingBoxMethod))
