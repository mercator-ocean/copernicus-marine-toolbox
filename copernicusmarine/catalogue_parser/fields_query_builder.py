from typing import Optional, Type, get_args, get_origin, get_type_hints

from pydantic import BaseModel


def check_type_is_base_model(type_to_check: Type) -> bool:
    try:
        return issubclass(type_to_check, BaseModel)
    except TypeError:
        return False


class QueryBuilder:
    fields_to_include_or_exclude: set[str]

    def __init__(self, fields_to_include_or_exclude: set[str]):
        self.fields_to_include_or_exclude = fields_to_include_or_exclude
        assert (
            self.fields_to_include_or_exclude
        ), "fields_to_include_or_exclude is empty"

    def build_query(
        self,
        type_to_check: Type,
        query: Optional[dict] = None,
    ):
        """
        Recursively builds a query to include or exclude fields from a Pydantic model
        Allows to handle nested models, lists and dictionaries.

        Example of resulting query:
        input:
        QueryBuilder({"product_id", "dataset_id"}).build_query(CopernicusMarineCatalogue)

        output:
        {
            "products": {
                "__all__": {
                    "product_id": True
                    "datasets": {
                        "__all__": {
                            "dataset_id": True
                        }
                    }
                },
            }
        }
        """  # noqa: E501
        if query is None:
            query = {}

        for (
            field_name,
            field_type,
        ) in get_type_hints(type_to_check).items():
            if field_name in self.fields_to_include_or_exclude:
                query[field_name] = True
            elif get_origin(field_type) is list:
                if field_name not in query:
                    query[field_name] = {"__all__": {}}
                result = self.build_query(
                    get_args(field_type)[0],
                    query[field_name]["__all__"],
                )
                if not result:
                    del query[field_name]
            elif get_origin(field_type) is dict:
                if field_name not in query:
                    query[field_name] = {"__all__": {}}
                result = self.build_query(
                    get_args(field_type)[1],
                    query[field_name]["__all__"],
                )
                if not result:
                    del query[field_name]
            elif check_type_is_base_model(field_type):
                if field_name not in query:
                    query[field_name] = {}
                result = self.build_query(field_type, query[field_name])
                if not result:
                    del query[field_name]

        return query
