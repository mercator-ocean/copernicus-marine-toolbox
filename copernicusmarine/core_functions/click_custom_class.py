import inspect
import logging

import click

from copernicusmarine.core_functions.deprecated_options import (
    log_deprecated_message,
)

logger = logging.getLogger("copernicusmarine")


class DeprecatedClickOption(click.Option):
    def __init__(self, *args, **kwargs):
        self.deprecated = kwargs.pop("deprecated", ())
        self.preferred = kwargs.pop("preferred", None)
        super().__init__(*args, **kwargs)


class CustomClickOptionsCommand(click.Command):
    def make_parser(self, ctx):
        parser = super().make_parser(ctx)

        # get the parser options
        options = set(parser._short_opt.values())
        options |= set(parser._long_opt.values())

        for option in options:
            if not isinstance(option.obj, DeprecatedClickOption):
                continue

            def make_process(an_option):
                orig_process = an_option.process
                deprecated = getattr(an_option.obj, "deprecated", None)
                preferred = getattr(an_option.obj, "preferred", None)
                msg = "Expected `deprecated` value for `{}`"
                assert deprecated is not None, msg.format(an_option.obj.name)

                def process(value, state):
                    frame = inspect.currentframe()
                    try:
                        if frame and frame.f_back:
                            opt = frame.f_back.f_locals.get("opt")
                    finally:
                        del frame

                    if opt in deprecated:
                        log_deprecated_message(opt, preferred)
                    return orig_process(value, state)

                return process

            option.process = make_process(option)

        return parser

    def format_epilog(self, ctx, formatter):
        if self.epilog:
            formatter.write_paragraph()
            for line in self.epilog.split("\n"):
                if ".. code-block::" in line:
                    continue
                formatter.write(line + "\n")
