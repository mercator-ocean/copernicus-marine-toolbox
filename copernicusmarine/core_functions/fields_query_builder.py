import logging
from typing import (
    Literal,
    Optional,
    Type,
    Union,
    get_args,
    get_origin,
    get_type_hints,
)

from pydantic import BaseModel

logger = logging.getLogger("copernicusmarine")


def check_type_is_base_model(type_to_check: Type) -> bool:
    try:
        return issubclass(type_to_check, BaseModel)
    except TypeError:
        return False


class WrongFieldsError(Exception):
    """
    Exception raised when the fields requested are not queryable.

    Please verify the fields requested are a field name in the response type model
    for the command you are using. See the response type page of `the documentation <https://toolbox-docs.marine.copernicus.eu/>`_ for more information.
    """  # noqa: E501

    def __init__(
        self,
        wrong_fields: set[str],
        available_fields: set[str],
        option_name: str,
        *args: object,
    ) -> None:
        super().__init__(*args)
        self.wrong_fields = wrong_fields
        self.available_fields = available_fields
        self.option_name = option_name

    def __str__(self) -> str:
        return (
            f"All ``{self.option_name}`` fields are invalid"
            f": {', '.join(sorted(self.wrong_fields))}.\n"
            f"Available fields: {', '.join(sorted(self.available_fields))}"
        )


def build_query(
    fields_to_include_or_exclude: set[str],
    type_to_check: Type,
    query: Optional[dict] = None,
) -> dict:
    """
    Recursively builds a query to include or exclude fields from a Pydantic model
    Allows to handle nested models, lists and dictionaries.

    Example of resulting query:
    input:
    build_query({"product_id", "dataset_id"}, CopernicusMarineCatalogue)

    output:
    {
        "products": {
            "__all__": {
                "product_id": True
                "datasets": {
                    "__all__": {
                        "dataset_id": True
                    }
                }
            },
        }
    }
    """  # noqa: E501
    if query is None:
        query = {}

    for (
        field_name,
        field_type,
    ) in get_type_hints(type_to_check).items():
        if field_name in fields_to_include_or_exclude:
            query[field_name] = True
            continue
        all_base_models = _get_base_models_in_type(field_type)
        for base_model, in_an_iterable in (all_base_models or {}).items():
            if field_name not in query:
                query[field_name] = {"__all__": {}} if in_an_iterable else {}
            result = build_query(
                fields_to_include_or_exclude,
                base_model,
                (
                    query[field_name]["__all__"]
                    if in_an_iterable
                    else query[field_name]
                ),
            )
            if not result:
                del query[field_name]
    return query


def _get_base_models_in_type(
    type_to_check: Type,
    in_an_iterable: bool = False,
) -> Optional[dict[Type, Literal["__all__", None]]]:
    models = {}
    if check_type_is_base_model(type_to_check):
        return {type_to_check: "__all__" if in_an_iterable else None}
    elif get_origin(type_to_check) is list:
        result = _get_base_models_in_type(
            get_args(type_to_check)[0], in_an_iterable=True
        )
        if result:
            models.update(result)
    elif get_origin(type_to_check) is dict:
        result = _get_base_models_in_type(
            get_args(type_to_check)[1], in_an_iterable=True
        )
        if result:
            models.update(result)
    elif get_origin(type_to_check) is Union:
        for union_type in get_args(type_to_check):
            result = _get_base_models_in_type(
                union_type, in_an_iterable=in_an_iterable
            )
            if result:
                models.update(result)
    return models


def _return_available_fields(
    type_to_check: Type, available_fields: Optional[set[str]] = None
) -> set[str]:
    """
    Reccursively get all the fields that are available in the model
    """
    if available_fields is None:
        available_fields = set()

    for (
        field_name,
        field_type,
    ) in get_type_hints(type_to_check).items():
        available_fields.add(field_name)
        all_base_models = _get_base_models_in_type(field_type)
        for base_model, _ in (all_base_models or {}).items():
            result = _return_available_fields(base_model, available_fields)
            available_fields.union(result)
    return available_fields


def get_queryable_requested_fields(
    requested_fields: set[str], type_to_check: Type, option_name: str
) -> set[str]:
    """
    Computes all the fields that are available in the model and checks if the requested fields are queryable

    if none of the fields are queryable, it raises an error
    if some of the fields are not queryable, it returns the queryable fields and logs a warning
    Returns:
    - a set with the queryable fields
    """  # noqa: E501
    available_fields = _return_available_fields(type_to_check)
    queryable_fields = requested_fields.intersection(available_fields)
    wrong_inputs = requested_fields - queryable_fields
    if queryable_fields == set():
        raise WrongFieldsError(wrong_inputs, available_fields, option_name)
    elif wrong_inputs:
        logger.warning(
            f"Some ``{option_name}`` fields are invalid:"
            f" {', '.join(wrong_inputs)}.\n "
            f"Available fields: {', '.join(sorted(available_fields))}."
        )
    return queryable_fields
