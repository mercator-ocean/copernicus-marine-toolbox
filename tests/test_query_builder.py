from typing import Literal

from copernicusmarine.catalogue_parser.models import CopernicusMarineCatalogue
from copernicusmarine.core_functions.fields_query_builder import (
    _return_available_fields,
    build_query,
)
from copernicusmarine.core_functions.models import ResponseGet, ResponseSubset


class TestQueryBuilder:
    # For the catalogue
    def test_query_builder_catalogue(self, snapshot):
        query = build_query(
            ({"product_id", "dataset_id"}), CopernicusMarineCatalogue
        )
        assert query == snapshot

    def test_query_builder_catalogue_deep_inside(self, snapshot):
        query = build_query(
            ({"service_name", "coordinate_unit", "maximum_value"}),
            CopernicusMarineCatalogue,
        )
        assert query == snapshot

    # For the get
    def test_query_builder_responseget(self, snapshot):
        query = build_query(({"files", "files_deleted"}), ResponseGet)
        assert query == snapshot

    def test_query_builder_responseget_inside_basemodel(self, snapshot):
        query = build_query(({"file_status"}), ResponseGet)
        assert query == snapshot

    # For the subset
    def test_basic_query_builder_responsesubset(self, snapshot):
        query = build_query(
            ({"file_path", "output_directory"}), ResponseSubset
        )
        assert query == snapshot

    def test_subset_optional_float(self, snapshot):
        query = build_query(
            ({"file_size", "data_transfer_size"}), ResponseSubset
        )
        assert query == snapshot

    def test_subset_optional_status(self, snapshot):
        query = build_query(({"status", "file_status"}), ResponseSubset)
        assert query == snapshot

    def test_subset_optional_coordinate_id(self, snapshot):
        query = build_query(({"coordinate_id"}), ResponseSubset)
        assert query == snapshot

    def test_subset_optional_coordinate_id_and_file(self, snapshot):
        query = build_query(({"coordinate_id", "file_status"}), ResponseSubset)
        assert query == snapshot

    def test_return_available_fields(self, snapshot):
        self.when_I_run_with_model("describe", snapshot)
        self.when_I_run_with_model("get", snapshot)
        self.when_I_run_with_model("subset", snapshot)

    def when_I_run_with_model(
        self, command: Literal["get", "describe", "subset"], snapshot
    ):
        if command == "describe":
            model = CopernicusMarineCatalogue
        else:
            if command == "get":
                model = ResponseGet
            else:
                model = ResponseSubset

        available_fields = _return_available_fields(model)

        assert sorted(list(available_fields)) == snapshot
