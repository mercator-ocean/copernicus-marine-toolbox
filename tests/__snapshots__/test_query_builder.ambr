# serializer version: 1
# name: TestQueryBuilder.test_basic_query_builder_responsesubset
  dict({
    'file_path': True,
    'output_directory': True,
  })
# ---
# name: TestQueryBuilder.test_query_builder_catalogue
  dict({
    'products': dict({
      '__all__': dict({
        'datasets': dict({
          '__all__': dict({
            'dataset_id': True,
          }),
        }),
        'product_id': True,
      }),
    }),
  })
# ---
# name: TestQueryBuilder.test_query_builder_catalogue_deep_inside
  dict({
    'products': dict({
      '__all__': dict({
        'datasets': dict({
          '__all__': dict({
            'versions': dict({
              '__all__': dict({
                'parts': dict({
                  '__all__': dict({
                    'services': dict({
                      '__all__': dict({
                        'service_name': True,
                        'variables': dict({
                          '__all__': dict({
                            'coordinates': dict({
                              '__all__': dict({
                                'coordinate_unit': True,
                                'maximum_value': True,
                              }),
                            }),
                          }),
                        }),
                      }),
                    }),
                  }),
                }),
              }),
            }),
          }),
        }),
      }),
    }),
  })
# ---
# name: TestQueryBuilder.test_query_builder_responseget
  dict({
    'files': True,
    'files_deleted': True,
  })
# ---
# name: TestQueryBuilder.test_query_builder_responseget_inside_basemodel
  dict({
    'files': dict({
      '__all__': dict({
        'file_status': True,
      }),
    }),
  })
# ---
# name: TestQueryBuilder.test_return_available_fields
  list([
    'arco_updated_date',
    'arco_updating_start_date',
    'axis',
    'bbox',
    'chunk_geometric_factor',
    'chunk_reference_coordinate',
    'chunk_type',
    'chunking_length',
    'coordinate_id',
    'coordinate_unit',
    'coordinates',
    'dataset_id',
    'dataset_name',
    'datasets',
    'description',
    'digital_object_identifier',
    'keywords',
    'label',
    'maximum_value',
    'minimum_value',
    'name',
    'parts',
    'platforms_metadata',
    'processing_level',
    'product_id',
    'production_center',
    'products',
    'released_date',
    'retired_date',
    'service_format',
    'service_name',
    'service_short_name',
    'services',
    'short_name',
    'sources',
    'standard_name',
    'step',
    'thumbnail_url',
    'title',
    'units',
    'uri',
    'url_metadata',
    'values',
    'variables',
    'versions',
  ])
# ---
# name: TestQueryBuilder.test_return_available_fields.1
  list([
    'etag',
    'file_format',
    'file_path',
    'file_size',
    'file_status',
    'filename',
    'files',
    'files_deleted',
    'files_not_found',
    'https_url',
    'last_modified_datetime',
    'message',
    'number_of_files_to_download',
    'output_directory',
    's3_url',
    'status',
    'total_size',
  ])
# ---
# name: TestQueryBuilder.test_return_available_fields.2
  list([
    'coordinate_id',
    'coordinates_extent',
    'data_transfer_size',
    'file_path',
    'file_size',
    'file_status',
    'filename',
    'maximum',
    'message',
    'minimum',
    'output_directory',
    'status',
    'unit',
    'variables',
  ])
# ---
# name: TestQueryBuilder.test_subset_optional_coordinate_id
  dict({
    'coordinates_extent': dict({
      '__all__': dict({
        'coordinate_id': True,
      }),
    }),
  })
# ---
# name: TestQueryBuilder.test_subset_optional_coordinate_id_and_file
  dict({
    'coordinates_extent': dict({
      '__all__': dict({
        'coordinate_id': True,
      }),
    }),
    'file_status': True,
  })
# ---
# name: TestQueryBuilder.test_subset_optional_float
  dict({
    'data_transfer_size': True,
    'file_size': True,
  })
# ---
# name: TestQueryBuilder.test_subset_optional_status
  dict({
    'file_status': True,
    'status': True,
  })
# ---
